c2rust-analyze/src/known_fn.rs

```diff
@@ -602,28 +622,823 @@ pub const fn all_known_fns() -> &'static [KnownFn] {
+            ) -> *mut tm: [READ | WRITE];
+
+            fn lseek(
+                fd: c_int,
+                offset: off_t,
+                whence: c_int,
+            ) -> off_t;
+
+            fn lstat(
+                path: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+                buf: *mut stat: [WRITE | NON_NULL],
+            ) -> c_int;
+
+            fn malloc(
+                size: size_t,
+            ) -> *mut c_void: [READ | WRITE | OFFSET_ADD | FREE]; // TODO(kkysen) OFFSET_ADD?
+
+            fn memchr(
+                // `WRITE` because the return type is derived from `cx`'s provenance.
+                cx: *const c_void: [READ | WRITE | OFFSET_ADD | NON_NULL],
+                c: c_int,
+                n: size_t,
+            ) -> *mut c_void: [READ | WRITE | OFFSET_ADD | OFFSET_SUB];
+
+            fn memcmp(
+                cx: *const c_void: [READ | OFFSET_ADD | NON_NULL],
+                ct: *const c_void: [READ | OFFSET_ADD | NON_NULL],
+                n: size_t,
+            ) -> c_int;
+
+            fn memcpy(
+                dest: *mut c_void: [WRITE | OFFSET_ADD | NON_NULL],
+                src: *const c_void: [READ | OFFSET_ADD | NON_NULL],
+                n: size_t,
+                // Same as `dest`, but while `dest` shouldn't be `NULL`,
+                // it can without error, and thus this can return `NULL`.
+            ) -> *mut c_void: [WRITE | OFFSET_ADD];
+
+            fn memmove(
+                // TODO(kkysen) Note: A `&mut []` for `dest` and `&[]` for `src` would be UB since they can overlap.
+                dest: *mut c_void: [WRITE | OFFSET_ADD | NON_NULL],
+                src: *const c_void: [READ | OFFSET_ADD | NON_NULL],
+                n: size_t,
+                // Same as `dest`, but while `dest` shouldn't be `NULL`,
+                // it can without error, and thus this can return `NULL`.
+            ) -> *mut c_void: [WRITE | OFFSET_ADD];
+
+            // TODO(kkysen) Not in `libc` crate.
+            // fn mempcpy;
+
+            fn memset(
+                dest: *mut c_void: [WRITE | OFFSET_ADD | NON_NULL],
+                c: c_int,
+                n: size_t,
+                // Same as `dest`, but while `dest` shouldn't be `NULL`,
+                // it can without error, and thus this can return `NULL`.
+            ) -> *mut c_void: [WRITE | OFFSET_ADD];
+
+            #[cfg(target_os = "linux")]
+            fn mkostemp(
+                template: *mut c_char: [READ | WRITE | OFFSET_ADD | NON_NULL],
+                flags: c_int,
+            ) -> c_int;
+
+            fn mmap(
+                // not yet a valid pointer
+                addr: *mut c_void: [NONE],
+                len: size_t,
+                prot: c_int,
+                flags: c_int,
+                fd: c_int,
+                offset: off_t,
+                // TODO(kkysen) not always a pointer, can be -1
+            ) -> *mut c_void: [READ | WRITE | OFFSET_ADD | NON_NULL];
```

Comment:
Unfortunate.  There's discussion somewhere of a `VALID` permission for handling cases like this, but of course it's not implemented at all yet.

Reply:
Yeah, it's unfortunate.  I'll leave the `TODO` and if we eventually add a `VALID` permission, we can handle it then.